src/tcpinfo_lib.cc

```diff
@@ -14,8 +14,207 @@
+  if (bbr->bbr_pacing_gain > 0) proto->set_pacing_gain(bbr->bbr_pacing_gain);
+  if (bbr->bbr_cwnd_gain > 0) proto->set_cwnd_gain(bbr->bbr_cwnd_gain);
+}
+
+#define TCPI_HAS_OPT(info, opt) !!(info->tcpi_options & (opt))
+
+void ParseTCPInfo(const struct rtattr* tb, TCPInfoProto* proto) {
+  const struct tcp_info* info;
+  int len = RTA_PAYLOAD(tb);
+
+  /* workaround for older kernels with fewer fields */
+  if (len < sizeof(*info)) {
+    info = (const struct tcp_info*)alloca(sizeof(*info));
+    memcpy((char *)info, (char*)RTA_DATA(tb), len);
+    memset((char *)info + len, 0, sizeof(*info) - len);
+  } else {
+    info = (const struct tcp_info*)RTA_DATA(tb);
+  }
+
+#define SET_NONZERO(field) \
+  if (info->tcpi_##field) \
+    proto->set_##field(info->tcpi_##field)
+#define SET_ALWAYS(field) \
+    proto->set_##field(info->tcpi_##field)
+
+  proto->set_state(TCPState(info->tcpi_state));  // error checking?
+  SET_ALWAYS(ca_state);
+  SET_NONZERO(retransmits);
+  SET_NONZERO(probes);
+  SET_NONZERO(backoff);
+  SET_ALWAYS(options);
+  // congestion?
+  if (TCPI_HAS_OPT(info, TCPI_OPT_WSCALE)) {
+    proto->set_snd_wscale(info->tcpi_snd_wscale);
+    proto->set_rcv_wscale(info->tcpi_rcv_wscale);
+  }
+  // delivery_rate_app_limited ...
+  //
+  if (info->tcpi_rto && info->tcpi_rto != 3000000)
+    proto->set_rto(info->tcpi_rto);
+  SET_NONZERO(ato);
+  SET_NONZERO(ato);
+  SET_NONZERO(snd_mss);
+  SET_NONZERO(rcv_mss);
+
+  SET_NONZERO(unacked);
+  SET_NONZERO(sacked);
+  SET_NONZERO(lost);
+  SET_NONZERO(retrans);
+  SET_NONZERO(fackets);  // ???
+
+  SET_ALWAYS(last_data_sent);
+  SET_ALWAYS(last_ack_sent);
+  SET_ALWAYS(last_data_recv);
+  SET_ALWAYS(last_ack_recv);
+
+  SET_NONZERO(pmtu);
+  if (info->tcpi_rcv_ssthresh < 0xFFFF)
+    proto->set_rcv_ssthresh(info->tcpi_rcv_ssthresh);
+  SET_NONZERO(rtt);
+  SET_NONZERO(rttvar);
+  SET_NONZERO(snd_ssthresh);
+  SET_NONZERO(snd_cwnd);
+  SET_NONZERO(advmss);
+  SET_NONZERO(reordering);
+
+  SET_NONZERO(rcv_rtt);
+  SET_NONZERO(rcv_space);
+  SET_NONZERO(total_retrans);
+
+  SET_NONZERO(pacing_rate);
+  SET_NONZERO(max_pacing_rate);
+  SET_NONZERO(bytes_acked);
+  SET_NONZERO(bytes_received);
+
+  SET_NONZERO(segs_out);
+  SET_NONZERO(segs_in);
+
+  SET_NONZERO(notsent_bytes);
+  SET_NONZERO(min_rtt);
+  SET_NONZERO(data_segs_in);
+  SET_NONZERO(data_segs_out);
+
+  SET_NONZERO(delivery_rate);
+#undef SET_NONZERO
+#undef SET_ALWAYS
+}
+
+void ParseMemInfo(const struct rtattr* tb,
+                  SocketMemInfoProto* proto) {
+  const auto* info = (const struct inet_diag_meminfo*)RTA_DATA(tb);
+  proto->set_rmem_alloc(info->idiag_rmem);
+  proto->set_wmem_alloc(info->idiag_wmem);
+  proto->set_fwd_alloc(info->idiag_fmem);
+  proto->set_tmem(info->idiag_tmem);
+}
+
+void ParseSKMemInfo(const struct rtattr* tb,
+                  SocketMemInfoProto* proto) {
+  const __u32* info = (__u32*)RTA_DATA(tb);
+#define SET_NONZERO(tag, field) \
```

Comment:
Please choose a different name? I recognize that you engaged in best practices and undefined the macro above, but it's still weird to see a macro have two different meanings in so short a time.